NuGet/NuGet.Services.EndToEnd
# Add initial end-to-end tests and polling test infrastructure

1. Add basic end-to-end tests which:
    1. Push a package and wait for it in flat container
    1. Push a package and wait for it in registration
    1. Push a package and wait for it in search
1. Rename `BasicTests` to `ConnectivityTests`
1. Add test infrastructure for:
    1. Sharing a pushed package between tests.
    1. Polling V3 endpoints for package availability
    1. Determining search staging URL (so we can end-to-end test search service before VIP swap). Not that we already have this for gallery.

I decided to implement the polling logic with simple HTTP requests and JSON parsing and no other dependencies. The alternative would be to use client-side APIs. I decided against client side APIs since I wanted a) minimal validation to be done during the polling, b) there to be no possibility of caching, and c) fast polling to minimize delays. a) and b) are both a little tricky with client APIs.

As we add more end-to-end tests, we can use these polling primitives during the "arrange" step of the tests and have more in-depth validations in the "assert" step. In other words, the polling should be considered as a setup step, not a validation step, for future end-to-end tests. Therefore, client APIs (which intrinsically validate a lot of things) are not ideal. 

In the future, we can decide to use more or less client APIs in our end-to-end tests.

So far, there is only one package that gets pushed per end-to-end test run. An example of the ID + version is:
```
E2E.SemVer1.170425.165504.0389598 1.0.0
```

As we add more packages, I am considering just pushing them all at once instead of on-demand/as tests need them. Thoughts on this?

End-to-end tests are configured run during the following deployments:
1. gallery (before VIP swap)
1. search (before VIP swap)
1. feed2catalog (after deployment)
1. catalog2registration (after deployment)
1. catalog2dnx (after deployment)
1. catalog2lucene (after deployment)

This PR is a prerequisite for https://github.com/NuGet/NuGetGallery/issues/3786.
